core/trie/proof.go

```diff
@@ -133,6 +133,213 @@ func transformNode(tri *Trie, parentKey *Key, sNode StorageNode) (*Edge, *Binary
+				return err
+			}
+		}
+	} else if node.Edge != nil {
+		edgeNode, exist := nodeHashes[*node.Edge.Child]
+		if exist {
+			if err := noCycleCheck(edgeNode, nodeHashes, visitedHashes, hash); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func rootNodeExistsCheck(rootHash *felt.Felt, nodeHashes map[felt.Felt]ProofNode) (ProofNode, error) {
+	currNode, rootExists := nodeHashes[*rootHash]
+	if !rootExists {
+		return currNode, errors.New("root hash not found in the merged path")
+	}
+
+	return currNode, nil
+}
+
+// traverseNodes traverses the merged proof path starting at `currNode`
+// and adds nodes to `path` slice. It stops when the split node is added
+// or the path is exhausted, and `currNode` children are not included
+// in the path (nodeHashes)
+func traverseNodes(currNode ProofNode, path *[]ProofNode, nodeHashes map[felt.Felt]ProofNode) {
+	*path = append(*path, currNode)
+
+	if currNode.Binary != nil {
+		expectedLeftHash := currNode.Binary.LeftHash
+		expectedRightHash := currNode.Binary.RightHash
+
+		nodeLeft, leftExist := nodeHashes[*expectedLeftHash]
+		nodeRight, rightExist := nodeHashes[*expectedRightHash]
+		if leftExist && rightExist {
+			return
+		} else if leftExist {
+			traverseNodes(nodeLeft, path, nodeHashes)
+		} else if rightExist {
+			traverseNodes(nodeRight, path, nodeHashes)
+		}
+	} else if currNode.Edge != nil {
+		edgeNode, exist := nodeHashes[*currNode.Edge.Child]
+		if exist {
+			traverseNodes(edgeNode, path, nodeHashes)
+		}
+	}
+}
+
+// MergeProofPaths removes duplicates and merges proof paths into a single path
+// merges the paths in the specified order [commonNodes..., leftNode, rightNode, leftNode...]
+// ordering of the merged path is not important
+// since splitProofPath can discover the left and right paths using the merged path and the rootHash
+func MergeProofPaths(leftPath, rightPath []ProofNode, hash hashFunc) ([]ProofNode, *felt.Felt, error) {
+	merged := []ProofNode{}
+	minLen := min(len(leftPath), len(rightPath))
+
+	if len(leftPath) == 0 || len(rightPath) == 0 {
+		return merged, nil, errors.New("empty proof paths")
+	}
+
+	if !leftPath[0].Hash(hash).Equal(rightPath[0].Hash(hash)) {
+		return merged, nil, errors.New("roots of the proof paths are different")
+	}
+
+	rootHash := leftPath[0].Hash(hash)
+
+	// Get duplicates and insert by one
+	i := 0
+	for i = 0; i < minLen; i++ {
+		leftNode := leftPath[i]
+		rightNode := rightPath[i]
+
+		if leftNode.Hash(hash).Equal(rightNode.Hash(hash)) {
+			merged = append(merged, leftNode)
+		} else {
+			break
+		}
+	}
+
+	// Add rest of the nodes one from left and one from right
+	// until we reach the end of the shortest path
+	for ; i < minLen; i++ {
+		merged = append(merged, leftPath[i], rightPath[i])
+	}
+
+	// Add the rest of the nodes from the longest path
+	if len(leftPath) > minLen {
+		merged = append(merged, leftPath[i:]...)
+	} else if len(rightPath) > minLen {
+		merged = append(merged, rightPath[i:]...)
+	}
```

Comment:
can we simplify to this?
merged = append(merged, leftPath[i:]...)
merged = append(merged, rightPath[i:]...)

Reply:
Yes it was trying to follow a format since we are not using this format anymore and ordering is not important in splitProof we can simplify this